source/unified-test-format/unified-test-format.rst

```diff
@@ -0,0 +1,1461 @@
+SHOULD always be written using the latest version of this specification.
+
+
+Entity Map
+----------
+
+The entity map indexes arbitrary objects and values by unique names, so that
+they can be referenced from test constructs (e.g.
+`operation.object <operation_object_>`_). To ensure each test is executed in
+isolation, test runners MUST NOT share entity maps between tests. Most entities
+will be driver objects created by the `createEntities`_ directive during test
+setup, but the entity map may also be modified during test execution via the
+`operation.saveResultAsEntity <operation_saveResultAsEntity_>`_ directive.
+
+Test runners may choose to implement the entity map in a fashion most suited to
+their language, but implementations MUST enforce both uniqueness of entity names
+and referential integrity when fetching an entity. Test runners MUST raise an
+error if an attempt is made to store an entity with a name that already exists
+in the map and MUST raise an error if an entity is not found for a name or is
+found but has an unexpected type.
+
+Consider the following examples::
+
+    # Error due to a duplicate name (client0 was already defined)
+    createEntities:
+      - client: { id: client0 }
+      - client: { id: client0 }
+
+    # Error due to a missing entity (client1 is undefined)
+    createEntities:
+      - client: { id: client0 }
+      - session: { id: session0, client: client1 }
+
+    # Error due to an unexpected entity type (session instead of client)
+    createEntities:
+      - client: { id: client0 }
+      - session: { id: session0, client: client0 }
+      - session: { id: session1, client: session0 }
+
+
+Test Format
+-----------
+
+Each specification test file can define one or more tests, which inherit some
+top-level configuration (e.g. namespace, initial data). YAML and JSON test files
+are parsed as a document by the test runner. This section defines the top-level
+keys for that document and links to various sub-sections for definitions of
+nested structures (e.g. individual `test`_, `operation`_).
+
+Although test runners are free to process YAML or JSON files, YAML is the
+canonical format for writing tests. YAML files may be converted to JSON using a
+tool such as `js-yaml <https://github.com/nodeca/js-yaml>`__ .
+
+
+Top-level Fields
+~~~~~~~~~~~~~~~~
+
+The top-level fields of a test file are as follows:
+
+.. _schemaVersion:
+
+- ``schemaVersion``: Required string. Version of this specification to which the
+  test file complies. Test runners will use this to determine compatibility
+  (i.e. whether and how the test file will be interpreted). The format of this
+  string is defined in `Version String`_.
+
+.. _runOn:
+
+- ``runOn``: Optional array of documents. List of server version and/or topology
+  requirements for which the tests in this file can be run. These requirements
+  may be overridden on a per-test basis by `test.runOn <test_runOn_>`_. Test
```

Comment:
For now, we'll add language to make the test-level runOn more restrictive and allow a top-level runOn to skip the entire file. We can always revisit later if needed.

Reply:
This sounds fine to me, is there a plan to make some sort of meta test suite for the test format? One that could prove that an implementation indeed does not allow a test-level `runOn` that's more permissive than its parent?

Reply:
> is there a plan to make some sort of meta test suite for the test format?

I do plan to write tests to exercise some of the test format rules (e.g. a file with just `initialData` and an `outcome` assertion). Additionally, I would port over (or reimplement) one test from each spec from the original scope ([WRITING-4394](https://jira.mongodb.org/browse/WRITING-4394)) to ensure the format is sensible for that spec. For CRUD, this might be a giant test with tons of operations and definitely something that exercises `bulkWrite`. For transactions, I'd definitely want to use some of the special assertions and `withTransaction`.

> One that could prove that an implementation indeed does not allow a test-level runOn that's more permissive than its parent?

I don't see how we'd enforce this in the test runner if the goal of a top-level `runOn` is to just allow the entire file to be skipped without considering the contents of `tests`. My intention was to instruct the test runner to process each independently and we'd tell folks writing tests to not use more permissive `test.runOn` requirements.

If that doesn't sit well with you, then we can pivot to one of two choices for cases where both are specified:

* If `test.runOn` is specified, those requirements are checked _in addition to_ any top-level `runOn`.Perhaps this is the most clear rules, since we won't have to think about inheritance and overriding anything (similar reasons we decided to get rid of top-level `databaseName` and `collectionName` fields today).
* If `test.runOn` is specified, those requirements are checked _instead of_ any  top-level `runOn`. I don't really like this, as the requirements shown at the top of a test file couldn't necessarily be trusted.

Reply:
I moved this discussion into the new "Open Questions" section. Feel free to continue conversation here or start a new thread on that section.

Reply:
I'm fine to resolve this thread, how about you @prashantmital ?

Reply:
Consensus in the comment thread on the open question is to instruct test files to only use _more_ restrictive `test.runOn` requirements, so test runners can still decide to skip an entire file based on the top-level `runOn`. Prashant and Isabel are in that thread, so I'll resolve this.